_posts/2017-05-10-fastlane-travis-weekly-deploys.md

```diff
@@ -0,0 +1,249 @@
+
+## Ensuring signing will work.
+
+This one is a bit tricker, luckily I've already set up one of our apps to use [fastlane match][] and I can re-use that infrastructure. As it is a closed repo, Travis did not have access to clone the repo. I fixed this by creating an access token for a user with read-only access to our match-codesigning repo, then exposed this as a private environment variable in CI which the Matchfile uses. E.g.
+
+```ruby
+git_url "https://#{ENV['GITHUB_SUBMODULES_USER']}@github.com/artsy/mobile_fastlane_match"
+
+# Instead of 
+# git_url "https://github.com/artsy/mobile_fastlane_match"
+```
+
+That's what's great about building a DSL that which sits above a real programming language, you give users a lot of flexibility. I added a two fastlane lanes for code signing:
+
+```ruby
+# In case you need to update the signing profiles for this app
+lane :update_signing do
+  match(type: 'appstore')
+end
+
+# Used by CI, will not sneakily update (the CI only has read-only access to the repo anyway)
+lane :setup_signing do
+  match(type: 'appstore', readonly: true)
+end
+```
+
+## Creating the build and shipping it to Testflight
+
+This is handled by [fastlane gym][gym], again, I made a separate fastlane lane so it could be tested independently.
+
+```ruby
+# You can run `bundle exec fastlane build` to test the build process locally.
+lane :build do
+  gym workspace: 'emission/Example/Emission.xcworkspace',
+      configuration: 'Deploy',
+      scheme: 'Emission',
+      clean: true
+end
+```
+
+It uses a scheme for deploys, which prioritises using AppHub over a local React Native server. Gym handles a lot of CLI ugliness for us, and works well.
+
+Sending the app to Testflight involves a little bit of work:
+
+```ruby
+# Get the last 10 lines of the CHANGELOG for Testflight
+changelog = 'emission/CHANGELOG.md'
+upcoming_release_notes = File.read(changelog).lines[0...12].join("\n")
+
+# Log into iTunes connect, get the latest version of the app we shipped, and how many builds we've sent too
+Spaceship::Tunes.login(ENV['FASTLANE_USERNAME'], ENV['FASTLANE_PASSWORD'])
+app = Spaceship::Tunes::Application.find('net.artsy.Emission')
+latest_version = app.build_trains.keys.sort.last
+train = app.build_trains[latest_version]
+build_version = train.builds.count + 1
+```
+
+This lets the deploy process figure out what the latest release version is, and how many builds have shipped for that version. Then those can be used to set the build version and create a tag associated with it.
+
+```ruby
+# Ship to testflight
+pilot changelog: upcoming_release_notes
+
+# Tag releases
+add_git_tag tag: tag
+`git remote add http https://#{ENV["GITHUB_SUBMODULES_USER"]}@github.com/artsy/emission.git`
+`git push http #{tag}`
+```
+
+[fastlane pilot][] is send off the compiled build to Testflight, and then a tag noting the deploy is pushed to the main repo.
```

Comment:
These missing links are not missing, they are footnotes:

```
[This Link][]

[This Link]: https://thingy.com
```


[This Link][]

I'm done with inline links on long posts 👍 

[This Link]: https://thingy.com